go-peer/main.go

```diff
@@ -168,27 +167,29 @@ func main() {
 
 		opts = append(opts,
 			libp2p.Transport(ws.New, ws.WithTLSConfig(&tls.Config{Certificates: certs})),
-			libp2p.ListenAddrStrings("/ip4/0.0.0.0/tcp/0/ws", "/ip6/::/tcp/0/ws"),
+			libp2p.ListenAddrStrings("/ip4/0.0.0.0/tcp/0/ws"),
 		)
 	}
 
-	if len(announceAddrs) > 0 {
-		var announce []multiaddr.Multiaddr
-		for _, addr := range announceAddrs {
-			announce = append(announce, multiaddr.StringCast(addr))
+	addressFactory := func(addrs []multiaddr.Multiaddr) []multiaddr.Multiaddr {
+		addr := *announceAddr
+		if len(addr) != 0 {
+			// printErr("Announce address: %s\n", addr)
+			maddr := multiaddr.StringCast(addr)
+			addrs = append(addrs, maddr)
+			return addrs
 		}
-		opts = append(opts, libp2p.AddrsFactory(func([]multiaddr.Multiaddr) []multiaddr.Multiaddr {
-			return announce
-		}))
+		return addrs
 	}
 
 	opts = append(opts,
 		libp2p.Identity(privk),
 		libp2p.Transport(quicTransport.NewTransport),
 		libp2p.Transport(tcpTransport.NewTCPTransport),
 		libp2p.Transport(webtransport.New),
-		libp2p.ListenAddrStrings("/ip4/0.0.0.0/udp/9091/quic-v1", "/ip4/0.0.0.0/udp/9092/quic-v1/webtransport", "/ip4/0.0.0.0/tcp/9090"),
-		),
+		libp2p.ListenAddrStrings("/ip4/0.0.0.0/udp/9095/quic-v1", "/ip4/0.0.0.0/udp/9096/quic-v1/webtransport"),
+		libp2p.AddrsFactory(addressFactory),
```

Comment:
@marten-seemann I want to confirm AddrsFactory is doing what I expect. To get this working on EC2, I pass in a webtransport multiaddr as a cmd line argument. This addr has the EC2 instance's public IP and because that IP address doesn't exist on any interface inside the instance, I cannot add it as a ListenAddrString
Will adding that maddr to the AddrsFactory make it a listen address for this node?

Reply:
No it won't. `AddrsFactory` is an ugly API that just messes with the addresses returned by `AllAddrs`. With our address pipeline refactor, we'll finally be able to kill it.

If you listen on 0.0.0.0, your node should be listening on all available interfaces, even if it doesn't know the IP address of them. Identify will then allow the node to _discover_ what those addresses are. And with the v0.27.1 release, we'll be inferring the WebTransport address from the observation of QUIC addresses, if you're running both listeners on the same port (see comment above).